lib/collection/src/grouping/types.rs

```diff
@@ -0,0 +1,184 @@
+use std::collections::HashSet;
+use std::hash::Hash;
+
+use segment::types::{ExtendedPointId, PointGroup, ScoredPoint};
+use AggregatorError::{BadKeyType, KeyNotFound};
+
+#[derive(PartialEq, Debug)]
+pub(super) enum AggregatorError {
+    BadKeyType,
+    KeyNotFound,
+}
+#[derive(Debug, Clone)]
+pub(super) struct Group {
+    pub hits: Vec<HashablePoint>,
+    pub key: GroupKey,
+    pub group_by: String,
+}
+
+impl Group {
+    pub(super) fn hydrate_from(&mut self, set: &HashSet<HashablePoint>) {
+        self.hits.iter_mut().for_each(|hit| {
+            if let Some(point) = set.get(hit) {
+                hit.0.payload = point.0.payload.clone();
+                hit.0.vector = point.0.vector.clone();
+            }
+        });
+    }
+}
+
+impl From<Group> for PointGroup {
+    fn from(group: Group) -> Self {
+        let mut group_id = serde_json::Map::new();
+        group_id.insert(group.group_by, group.key.into());
+
+        Self {
+            hits: group.hits.into_iter().map(|hp| hp.0).collect(),
+            group_id,
+        }
+    }
+}
+
+/// Abstraction over serde_json::Value to be used as a key in a HashMap/HashSet
+#[derive(Debug, Eq, PartialEq, Clone)]
+pub(super) enum GroupKey {
+    String(String),
+    Number(serde_json::Number),
+}
+
+impl TryFrom<serde_json::Value> for GroupKey {
+    type Error = AggregatorError;
+
+    /// Only allows Strings and Numbers to be converted into GroupKey
+    /// When dealing with arrays, it will consider only the first element
+    fn try_from(value: serde_json::Value) -> Result<Self, Self::Error> {
+        let value = match value {
+            serde_json::Value::Array(arr) => arr.into_iter().next().ok_or(KeyNotFound)?,
+            _ => value,
+        };
+        match value {
+            serde_json::Value::String(s) => Ok(Self::String(s)),
+            serde_json::Value::Number(n) => Ok(Self::Number(n)),
+            _ => Err(BadKeyType),
+        }
+    }
+}
+
+impl From<GroupKey> for serde_json::Value {
+    fn from(key: GroupKey) -> Self {
+        match key {
+            GroupKey::String(s) => serde_json::Value::String(s),
+            GroupKey::Number(n) => serde_json::Value::Number(n),
+        }
+    }
+}
+
+impl Hash for GroupKey {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        match &self {
+            Self::Number(n) => n.hash(state),
+            Self::String(s) => s.hash(state),
+        }
+    }
+}
```

Comment:
`#[derive(Hash)]`?

```suggestion
/// Abstraction over serde_json::Value to be used as a key in a HashMap/HashSet
#[derive(Debug, Eq, PartialEq, Clone, Hash)]
pub(super) enum GroupKey {
    String(String),
    Number(serde_json::Number),
}

impl TryFrom<serde_json::Value> for GroupKey {
    type Error = AggregatorError;

    /// Only allows Strings and Numbers to be converted into GroupKey
    /// When dealing with arrays, it will consider only the first element
    fn try_from(value: serde_json::Value) -> Result<Self, Self::Error> {
        let value = match value {
            serde_json::Value::Array(arr) => arr.into_iter().next().ok_or(KeyNotFound)?,
            _ => value,
        };
        match value {
            serde_json::Value::String(s) => Ok(Self::String(s)),
            serde_json::Value::Number(n) => Ok(Self::Number(n)),
            _ => Err(BadKeyType),
        }
    }
}

impl From<GroupKey> for serde_json::Value {
    fn from(key: GroupKey) -> Self {
        match key {
            GroupKey::String(s) => serde_json::Value::String(s),
            GroupKey::Number(n) => serde_json::Value::Number(n),
        }
    }
}
```